In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn import metrics

# Data exploration

In [2]:
df1  = pd.read_csv(r'C:\Users\Naidu Vinay\Desktop\Astroid\dataset.csv')

C:\ProgramData\Anaconda3\lib\site-packages\IPython\core\interactiveshell.py:3444: DtypeWarning: Columns (3,4,5) have mixed types.Specify dtype option on import or set low_memory=False.
  exec(code_obj, self.user_global_ns, self.user_ns)


In [3]:
pd.set_option('display.max_columns', 500)

In [4]:
df1.head()

id    spkid       full_name pdes     name prefix neo pha     H  \
0  a0000001  2000001         1 Ceres    1    Ceres    NaN   N   N  3.40   
1  a0000002  2000002        2 Pallas    2   Pallas    NaN   N   N  4.20   
2  a0000003  2000003          3 Juno    3     Juno    NaN   N   N  5.33   
3  a0000004  2000004         4 Vesta    4    Vesta    NaN   N   N  3.00   
4  a0000005  2000005       5 Astraea    5  Astraea    NaN   N   N  6.90   

   diameter  albedo  diameter_sigma orbit_id      epoch  epoch_mjd  \
0   939.400  0.0900           0.200   JPL 47  2458600.5      58600   
1   545.000  0.1010          18.000   JPL 37  2459000.5      59000   
2   246.596  0.2140          10.594  JPL 112  2459000.5      59000   
3   525.400  0.4228           0.200   JPL 35  2458600.5      58600   
4   106.699  0.2740           3.140  JPL 114  2459000.5      59000   

    epoch_cal equinox         e         a         q          i          om  \
0  20190427.0   J2000  0.076009  2.769165  2.558684  10.594067   80.305531   
1  20200531.0   J2000  0.229972  2.773841  2.135935  34.832932  173.024741   
2  20200531.0   J2000  0.256936  2.668285  1.982706  12.991043  169.851482   
3  20190427.0   J2000  0.088721  2.361418  2.151909   7.141771  103.810804   
4  20200531.0   J2000  0.190913  2.574037  2.082619   5.367427  141.571026   

            w          ma        ad         n            tp        tp_cal  \
0   73.597695   77.372098  2.979647  0.213885  2.458239e+06  2.018043e+07   
1  310.202392  144.975675  3.411748  0.213345  2.458321e+06  2.018072e+07   
2  248.066193  125.435355  3.353865  0.226129  2.458446e+06  2.018112e+07   
3  150.728541   95.861938  2.570926  0.271609  2.458248e+06  2.018051e+07   
4  358.648418   17.846343  3.065455  0.238661  2.458926e+06  2.020032e+07   

           per     per_y     moid     moid_ld       sigma_e       sigma_a  \
0  1683.145703  4.608202  1.59478  620.640533  4.819000e-12  1.032800e-11   
1  1687.410992  4.619880  1.23429  480.348639  3.193400e-08  4.033700e-09   
2  1592.013769  4.358696  1.03429  402.514639  3.052000e-08  3.471800e-09   
3  1325.432763  3.628837  1.13948  443.451432  2.332100e-10  1.514300e-09   
4  1508.414421  4.129814  1.09575  426.433027  2.373700e-08  3.970900e-09   

        sigma_q       sigma_i      sigma_om       sigma_w      sigma_ma  \
0  1.956900e-11  4.608900e-09  6.168800e-08  6.624800e-08  7.820700e-09   
1  8.832200e-08  3.469400e-06  6.272400e-06  9.128200e-06  8.859100e-06   
2  8.139200e-08  3.223100e-06  1.664600e-05  1.772100e-05  8.110400e-06   
3  1.928600e-09  2.170600e-07  3.880800e-07  1.789300e-07  1.206800e-06   
4  6.092400e-08  2.740800e-06  2.894900e-05  2.984200e-05  8.303800e-06   

       sigma_ad       sigma_n      sigma_tp     sigma_per class      rms  
0  1.111300e-11  1.196500e-12  3.782900e-08  9.415900e-09   MBA  0.43301  
1  4.961300e-09  4.653600e-10  4.078700e-05  3.680700e-06   MBA  0.35936  
2  4.363900e-09  4.413400e-10  3.528800e-05  3.107200e-06   MBA  0.33848  
3  1.648600e-09  2.612500e-10  4.103700e-06  1.274900e-06   MBA  0.39980  
4  4.729000e-09  5.522700e-10  3.474300e-05  3.490500e-06   MBA  0.52191

In [5]:
df1.count()

id                958524
spkid             958524
full_name         958524
pdes              958524
name               22064
prefix                18
neo               958520
pha               938603
H                 952261
diameter          136209
albedo            135103
diameter_sigma    136081
orbit_id          958524
epoch             958524
epoch_mjd         958524
epoch_cal         958524
equinox           958524
e                 958524
a                 958524
q                 958524
i                 958524
om                958524
w                 958524
ma                958523
ad                958520
n                 958524
tp                958524
tp_cal            958524
per               958520
per_y             958523
moid              938603
moid_ld           958397
sigma_e           938602
sigma_a           938602
sigma_q           938602
sigma_i           938602
sigma_om          938602
sigma_w           938602
sigma_ma          938602
sigma_ad          938598


In [6]:
df1.dtypes


id                 object
spkid               int64
full_name          object
pdes               object
name               object
prefix             object
neo                object
pha                object
H                 float64
diameter          float64
albedo            float64
diameter_sigma    float64
orbit_id           object
epoch             float64
epoch_mjd           int64
epoch_cal         float64
equinox            object
e                 float64
a                 float64
q                 float64
i                 float64
om                float64
w                 float64
ma                float64
ad                float64
n                 float64
tp                float64
tp_cal            float64
per               float64
per_y             float64
moid              float64
moid_ld           float64
sigma_e           float64
sigma_a           float64
sigma_q           float64
sigma_i           float64
sigma_om          float64
sigma_w           float64
sigma_ma    

In [7]:
pd.set_option('display.max_columns', 500)
df1.describe()

spkid              H       diameter         albedo  \
count  9.585240e+05  952261.000000  136209.000000  135103.000000   
mean   3.810114e+06      16.906411       5.506429       0.130627   
std    6.831541e+06       1.790405       9.425164       0.110323   
min    2.000001e+06      -1.100000       0.002500       0.001000   
25%    2.239632e+06      16.100000       2.780000       0.053000   
50%    2.479262e+06      16.900000       3.972000       0.079000   
75%    3.752518e+06      17.714000       5.765000       0.190000   
max    5.401723e+07      33.200000     939.400000       1.000000   

       diameter_sigma         epoch      epoch_mjd     epoch_cal  \
count   136081.000000  9.585240e+05  958524.000000  9.585240e+05   
mean         0.479184  2.458869e+06   58868.781950  2.019693e+07   
std          0.782895  7.016716e+02     701.671573  1.930354e+04   
min          0.000500  2.425052e+06   25051.000000  1.927062e+07   
25%          0.180000  2.459000e+06   59000.000000  2.020053e+07   
50%          0.332000  2.459000e+06   59000.000000  2.020053e+07   
75%          0.620000  2.459000e+06   59000.000000  2.020053e+07   
max        140.000000  2.459000e+06   59000.000000  2.020053e+07   

                   e              a              q              i  \
count  958524.000000  958524.000000  958524.000000  958524.000000   
mean        0.156116       2.902143       2.394887       9.054901   
std         0.092643      39.719503       2.139246       6.637656   
min         0.000000  -14702.447872       0.070511       0.007744   
25%         0.092193       2.387835       1.971401       4.165419   
50%         0.145002       2.646969       2.225816       7.415665   
75%         0.200650       3.001932       2.578315      12.407550   
max         1.855356   33488.895955      80.398819     175.082901   

                  om              w             ma             ad  \
count  958524.000000  958524.000000  958523.000000  958520.000000   
mean      168.456416     181.369457     177.352582       3.455820   
std       102.881707     103.887995     105.816844      71.769401   
min         0.000025       0.000130     -70.703676       0.653773   
25%        80.604399      91.482063      83.785012       2.781603   
50%       159.826919     182.358081     175.641078       3.046995   
75%       252.247751     271.521589     269.865749       3.364793   
max       359.999793     359.999646     491.618014   66972.796064   

                  n            tp        tp_cal           per         per_y  \
count  9.585240e+05  9.585240e+05  9.585240e+05  9.585200e+05  9.585230e+05   
mean   2.366846e-01  2.458856e+06  2.019581e+07  5.085344e+03  1.392287e+01   
std    7.941705e-02  1.580506e+03  4.337669e+04  2.308961e+06  6.321581e+03   
min    1.608247e-07  2.283183e+06  1.539011e+07  1.511918e+02  0.000000e+00   
25%    1.894969e-01  2.458560e+06  2.019032e+07  1.347738e+03  3.689903e+00   
50%    2.288655e-01  2.458945e+06  2.020041e+07  1.572976e+03  4.306571e+00   
75%    2.671142e-01  2.459361e+06  2.021053e+07  1.899765e+03  5.201272e+00   
max    2.381082e+00  2.546362e+06  2.259081e+07  2.238462e+09  6.128574e+06   

               moid        moid_ld       sigma_e       sigma_a       sigma_q  \
count  9.386030e+05  958397.000000  9.386020e+05  9.386020e+05  9.386020e+05   
mean   1.415162e+00     539.364207  7.262721e-01  1.541424e+01  1.982929e+01   
std    2.156731e+00     834.306975  9.336391e+01  4.392381e+03  2.903785e+03   
min    4.544120e-07       0.000000  4.819000e-12  1.032800e-11  1.956900e-11   
25%    9.797020e-01     375.394939  5.474900e-08  2.046500e-08  1.462000e-07   
50%    1.240850e+00     477.706175  8.171600e-08  3.849150e-08  2.271900e-07   
75%    1.593370e+00     617.324804  2.336475e-07  1.044675e-07  6.583200e-07   
max    7.947660e+01   30929.908422  3.942500e+04  3.241200e+06  1.015000e+06   

            sigma_i      sigma_om       sigma_w      sigma_ma      sigma_ad  \
count  9.386020e+05  9.386020e+0

In [8]:
df1.shape

(958524, 45)

In [9]:
df1.columns


Index(['id', 'spkid', 'full_name', 'pdes', 'name', 'prefix', 'neo', 'pha', 'H',
       'diameter', 'albedo', 'diameter_sigma', 'orbit_id', 'epoch',
       'epoch_mjd', 'epoch_cal', 'equinox', 'e', 'a', 'q', 'i', 'om', 'w',
       'ma', 'ad', 'n', 'tp', 'tp_cal', 'per', 'per_y', 'moid', 'moid_ld',
       'sigma_e', 'sigma_a', 'sigma_q', 'sigma_i', 'sigma_om', 'sigma_w',
       'sigma_ma', 'sigma_ad', 'sigma_n', 'sigma_tp', 'sigma_per', 'class',
       'rms'],
      dtype='object')

Analyse Columns :



id and spkid

In [10]:
print(df1['id'].nunique())
print(df1['spkid'].nunique())
print(df1['full_name'].nunique())
print(df1['pdes'].nunique())
print(df1['name'].nunique())
print(df1['prefix'].nunique())


958524
958524
958524
958524
22064
1


Columns of type object

In [11]:
print(df1['pha'].unique())
print(df1['pha'].nunique())

['N' 'Y' nan]
2


In [12]:
print(df1['neo'].unique())
print(df1['neo'].nunique())


['N' 'Y' nan]
2


In [13]:
print(df1['orbit_id'].unique())
print(df1['orbit_id'].nunique())

['JPL 47' 'JPL 37' 'JPL 112' ... 'E2020H77' 'E2020K86' 'E2020J48']
4690


In [14]:
print(df1['equinox'].unique())
print(df1['equinox'].nunique())


['J2000']
1


In [15]:
print(df1['class'].unique())
print(df1['class'].nunique())

['MBA' 'OMB' 'MCA' 'AMO' 'IMB' 'TJN' 'CEN' 'APO' 'ATE' 'AST' 'TNO' 'IEO'
 'HYA']
13


In [16]:
df2 = df1.drop(['id', 'full_name', 'pdes', 'name', 'prefix', 'equinox'], axis = 'columns' )

# Data Wrangling

In [17]:
df2.isnull().mean()*100

spkid              0.000000
neo                0.000417
pha                2.078300
H                  0.653400
diameter          85.789714
albedo            85.905100
diameter_sigma    85.803068
orbit_id           0.000000
epoch              0.000000
epoch_mjd          0.000000
epoch_cal          0.000000
e                  0.000000
a                  0.000000
q                  0.000000
i                  0.000000
om                 0.000000
w                  0.000000
ma                 0.000104
ad                 0.000417
n                  0.000000
tp                 0.000000
tp_cal             0.000000
per                0.000417
per_y              0.000104
moid               2.078300
moid_ld            0.013250
sigma_e            2.078404
sigma_a            2.078404
sigma_q            2.078404
sigma_i            2.078404
sigma_om           2.078404
sigma_w            2.078404
sigma_ma           2.078404
sigma_ad           2.078821
sigma_n            2.078404
sigma_tp           2

In [18]:
df3 = df2.drop(['diameter', 'albedo', 'diameter_sigma'], axis = 'columns')

In [19]:
df4 = df3.dropna()

In [20]:
df4['neo'].value_counts(normalize = True)*100 #Percentage of asteriods as near earth objects.

N    97.545625
Y     2.454375
Name: neo, dtype: float64

In [21]:
df4[df4['neo'] == 'Y']['pha'].value_counts(normalize = True)*100 #Percentage of asteroids which are near earth objects which are potentially hazardous

N    90.971464
Y     9.028536
Name: pha, dtype: float64

In [22]:
df4[df4['pha'] == 'Y']['neo'].value_counts(normalize = True)*100 #Potentially hazardous ateroisa which are near earht objects.

Y    100.0
Name: neo, dtype: float64

In [23]:
df4['class'].value_counts(normalize = True)*100 #Distribution of orbit class of Ateroids

MBA    89.308028
OMB     2.914189
IMB     2.113189
MCA     1.908005
APO     1.360455
AMO     0.906112
TJN     0.871146
TNO     0.371004
ATE     0.185448
CEN     0.053951
AST     0.006114
IEO     0.002360
Name: class, dtype: float64

In [24]:
df4['orbit_id'].value_counts()

1          50142
JPL 1      47494
JPL 2      34563
JPL 3      29905
12         29136
           ...  
JPL 520        1
JPL 373        1
JPL 428        1
303            1
JPL 187        1
Name: orbit_id, Length: 525, dtype: int64

In [25]:
df4['orbit_id'].nunique() #Orbit IDs of Asteroids

525

# Data Preprocessing

In [26]:
df4.head()

spkid neo pha     H orbit_id      epoch  epoch_mjd   epoch_cal         e  \
0  2000001   N   N  3.40   JPL 47  2458600.5      58600  20190427.0  0.076009   
1  2000002   N   N  4.20   JPL 37  2459000.5      59000  20200531.0  0.229972   
2  2000003   N   N  5.33  JPL 112  2459000.5      59000  20200531.0  0.256936   
3  2000004   N   N  3.00   JPL 35  2458600.5      58600  20190427.0  0.088721   
4  2000005   N   N  6.90  JPL 114  2459000.5      59000  20200531.0  0.190913   

          a         q          i          om           w          ma  \
0  2.769165  2.558684  10.594067   80.305531   73.597695   77.372098   
1  2.773841  2.135935  34.832932  173.024741  310.202392  144.975675   
2  2.668285  1.982706  12.991043  169.851482  248.066193  125.435355   
3  2.361418  2.151909   7.141771  103.810804  150.728541   95.861938   
4  2.574037  2.082619   5.367427  141.571026  358.648418   17.846343   

         ad         n            tp        tp_cal          per     per_y  \
0  2.979647  0.213885  2.458239e+06  2.018043e+07  1683.145703  4.608202   
1  3.411748  0.213345  2.458321e+06  2.018072e+07  1687.410992  4.619880   
2  3.353865  0.226129  2.458446e+06  2.018112e+07  1592.013769  4.358696   
3  2.570926  0.271609  2.458248e+06  2.018051e+07  1325.432763  3.628837   
4  3.065455  0.238661  2.458926e+06  2.020032e+07  1508.414421  4.129814   

      moid     moid_ld       sigma_e       sigma_a       sigma_q  \
0  1.59478  620.640533  4.819000e-12  1.032800e-11  1.956900e-11   
1  1.23429  480.348639  3.193400e-08  4.033700e-09  8.832200e-08   
2  1.03429  402.514639  3.052000e-08  3.471800e-09  8.139200e-08   
3  1.13948  443.451432  2.332100e-10  1.514300e-09  1.928600e-09   
4  1.09575  426.433027  2.373700e-08  3.970900e-09  6.092400e-08   

        sigma_i      sigma_om       sigma_w      sigma_ma      sigma_ad  \
0  4.608900e-09  6.168800e-08  6.624800e-08  7.820700e-09  1.111300e-11   
1  3.469400e-06  6.272400e-06  9.128200e-06  8.859100e-06  4.961300e-09   
2  3.223100e-06  1.664600e-05  1.772100e-05  8.110400e-06  4.363900e-09   
3  2.170600e-07  3.880800e-07  1.789300e-07  1.206800e-06  1.648600e-09   
4  2.740800e-06  2.894900e-05  2.984200e-05  8.303800e-06  4.729000e-09   

        sigma_n      sigma_tp     sigma_per class      rms  
0  1.196500e-12  3.782900e-08  9.415900e-09   MBA  0.43301  
1  4.653600e-10  4.078700e-05  3.680700e-06   MBA  0.35936  
2  4.413400e-10  3.528800e-05  3.107200e-06   MBA  0.33848  
3  2.612500e-10  4.103700e-06  1.274900e-06   MBA  0.39980  
4  5.522700e-10  3.474300e-05  3.490500e-06   MBA  0.52191

Orbit Id has 525 different elements, we group Orbits less than 10 occurences as Other Ids

In [27]:
Orbit_Id_stats = df4.groupby('orbit_id')['orbit_id'].agg('count').sort_values(ascending = False)
Orbit_Id_stats

orbit_id
1          50142
JPL 1      47494
JPL 2      34563
JPL 3      29905
12         29136
           ...  
JPL 200        1
JPL 202        1
JPL 203        1
456            1
JPL 204        1
Name: orbit_id, Length: 525, dtype: int64

In [28]:
Orbit_Id_stats_lessthan10 = Orbit_Id_stats[Orbit_Id_stats<10]

In [29]:
Orbit_Id_stats_lessthan10

orbit_id
110        9
92         9
JPL 89     9
JPL 101    9
JPL 121    9
          ..
JPL 200    1
JPL 202    1
JPL 203    1
456        1
JPL 204    1
Name: orbit_id, Length: 331, dtype: int64

In [30]:
df5 = df4.copy() 
df5

spkid neo pha       H orbit_id      epoch  epoch_mjd   epoch_cal  \
0       2000001   N   N   3.400   JPL 47  2458600.5      58600  20190427.0   
1       2000002   N   N   4.200   JPL 37  2459000.5      59000  20200531.0   
2       2000003   N   N   5.330  JPL 112  2459000.5      59000  20200531.0   
3       2000004   N   N   3.000   JPL 35  2458600.5      58600  20190427.0   
4       2000005   N   N   6.900  JPL 114  2459000.5      59000  20200531.0   
...         ...  ..  ..     ...      ...        ...        ...         ...   
958519  3246801   N   N  17.135    JPL 5  2437203.5      37203  19600926.0   
958520  3246834   N   N  18.500        8  2459000.5      59000  20200531.0   
958521  3013075   Y   Y  20.400       17  2459000.5      59000  20200531.0   
958522  3246457   N   N  18.071    JPL 3  2441956.5      41956  19731001.0   
958523  3246553   N   N  18.060    JPL 3  2443429.5      43429  19771013.0   

               e         a         q          i          om           w  \
0       0.076009  2.769165  2.558684  10.594067   80.305531   73.597695   
1       0.229972  2.773841  2.135935  34.832932  173.024741  310.202392   
2       0.256936  2.668285  1.982706  12.991043  169.851482  248.066193   
3       0.088721  2.361418  2.151909   7.141771  103.810804  150.728541   
4       0.190913  2.574037  2.082619   5.367427  141.571026  358.648418   
...          ...       ...       ...        ...         ...         ...   
958519  0.185919  2.280861  1.856807   5.984416  194.229902  195.442802   
958520  0.282920  2.334910  1.674319   8.082280  355.248239    5.216223   
958521  0.662446  2.817152  0.950941   4.679278  182.982832  234.894852   
958522  0.202053  2.373137  1.893638   0.732484  176.499082  198.026527   
958523  0.258348  2.260404  1.676433   9.661947  204.512448  148.496988   

                ma        ad         n            tp        tp_cal  \
0        77.372098  2.979647  0.213885  2.458239e+06  2.018043e+07   
1       144.975675  3.411748  0.213345  2.458321e+06  2.018072e+07   
2       125.435355  3.353865  0.226129  2.458446e+06  2.018112e+07   
3        95.861938  2.570926  0.271609  2.458248e+06  2.018051e+07   
4        17.846343  3.065455  0.238661  2.458926e+06  2.020032e+07   
...            ...       ...       ...           ...           ...   
958519  342.057565  2.704916  0.286125  2.437266e+06  1.960113e+07   
958520  260.113042  2.995502  0.276248  2.459362e+06  2.021053e+07   
958521  240.473601  4.683363  0.208444  2.459574e+06  2.021123e+07   
958522  355.351127  2.852636  0.269600  2.441974e+06  1.973102e+07   
958523   15.320134  2.844376  0.290018  2.443377e+06  1.977082e+07   

                per     per_y      moid     moid_ld       sigma_e  \
0       1683.145703  4.608202  1.594780  620.640533  4.819000e-12   
1       1687.410992  4.619880  1.234290  480.348639  3.193400e-08   
2       1592.013769  4.358696  1.034290  402.514639  3.052000e-08   
3       1325.432763  3.628837  1.139480  443.451432  2.332100e-10   
4       1508.414421  4.129814  1.095750  426.433027  2.373700e-08   
...             ...       ...       ...         ...           ...   
958519  1258.191614  3.444741  0.861979  335.456367  2.142200e-02   
958520  1303.177916  3.567907  0.671242  261.227249  2.642000e-07   
958521  1727.085364  4.728502  0.035589   13.850093  6.636800e-08   
958522  1335.311579  3.655884  0.893896  347.877506  3.169800e-02   
958523  1241.302609  3.398501  0.680220  264.721217  1.486800e-02   

             sigma_a       sigma_q       sigma_i      sigma_om       sigma_w  \
0       1.032800e-11  1.956900e-11  4.608900e-09  6.168800e-08  6.624800e-08   
1       4.033700e-09  8.832200e-08  3.469400e-06  6.272400e-06  9.128200e-06   
2       3.471800e-09  8.139200e-08  3.223100e-06  1.664600e-05  1.772100e-05   
3       1.514300e-09  1.928600e-09  2.170600e-07  3.880800e-07  1.789300e-07   
4       3.970900e-09  6.092400e-08  2.740800e-06  2.894900e-05  2.984200e-05   
...              ...  

In [31]:
df5['orbit_id'] = df5['orbit_id'].apply(lambda x : 'Other_id' if x in Orbit_Id_stats_lessthan10 else x)

In [32]:
df5['orbit_id'].nunique()

195

In [33]:
df5.shape

(932335, 36)

In [34]:
df5['spkid'].nunique() #We have number of Id's equal to number of elements

932335

In [35]:
#df5.set_index('spkid') # We set the index equal to id's

Min MAx Scaling for Standardisation.

In [36]:
#Splitting the data to numeric and string
df6 = df5.copy()
df6.drop(['neo', 'pha', 'orbit_id', 'class'], axis = 'columns', inplace= True)
df7 = df5[['neo', 'orbit_id', 'class']]
df8 = df5['pha']
df8.shape
df8 = df8.to_numpy()

In [37]:
print(df6.shape,df7.shape,df8.shape)

(932335, 32) (932335, 3) (932335,)


In [38]:
# Normalisation using MinMax Scaler
from sklearn.preprocessing import MinMaxScaler
scalar = MinMaxScaler()
df9 = scalar.fit_transform(df6)
#df = pd.DataFrame(my_array, columns = ['Column_A','Column_B','Column_C'])
df9 = pd.DataFrame(df9, columns = df6.columns)

In [39]:
df9.shape

(932335, 32)

In [40]:
df9.head()

spkid         H     epoch  epoch_mjd  epoch_cal         e         a  \
0  0.000000e+00  0.131195  0.988218   0.988218   0.989134  0.076017  0.000066   
1  1.922440e-08  0.154519  1.000000   1.000000   1.000000  0.230004  0.000066   
2  3.844880e-08  0.187464  1.000000   1.000000   1.000000  0.256972  0.000063   
3  5.767320e-08  0.119534  0.988218   0.988218   0.989134  0.088732  0.000054   
4  7.689760e-08  0.233236  1.000000   1.000000   1.000000  0.190939  0.000060   

          q         i        om         w        ma        ad         n  \
0  0.030975  0.060467  0.223071  0.204438  0.258627  0.000035  0.089827   
1  0.025712  0.198916  0.480625  0.861674  0.379616  0.000041  0.089600   
2  0.023805  0.074158  0.471810  0.689073  0.344645  0.000040  0.094969   
3  0.025911  0.040748  0.288363  0.418691  0.291718  0.000029  0.114070   
4  0.025049  0.030614  0.393253  0.996247  0.152094  0.000036  0.100232   

         tp    tp_cal           per         per_y      moid   moid_ld  \
0  0.665157  0.665258  6.843780e-07  6.843780e-07  0.020066  0.020066   
1  0.665470  0.665298  6.862835e-07  6.862835e-07  0.015530  0.015530   
2  0.665944  0.665354  6.436662e-07  6.436662e-07  0.013014  0.013014   
3  0.665191  0.665269  5.245750e-07  5.245750e-07  0.014337  0.014337   
4  0.667768  0.668020  6.063194e-07  6.063194e-07  0.013787  0.013787   

        sigma_e       sigma_a       sigma_q       sigma_i      sigma_om  \
0  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
1  8.098714e-13  1.241322e-15  8.699747e-14  6.262048e-11  5.179478e-12   
2  7.740059e-13  1.067960e-15  8.016988e-14  5.816901e-11  1.383063e-11   
3  5.793050e-15  4.640170e-16  1.880819e-15  3.839709e-12  2.721975e-13   
4  6.019577e-13  1.221946e-15  6.000437e-14  4.945222e-11  2.409083e-11   

        sigma_w      sigma_ma      sigma_ad       sigma_n      sigma_tp  \
0  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
1  2.713159e-16  2.650639e-16  8.984495e-16  6.029037e-14  3.912546e-18   
2  5.285854e-16  2.426430e-16  7.900225e-16  5.717040e-14  3.384558e-18   
3  3.373713e-18  3.590511e-17  2.972008e-16  3.377845e-14  3.903861e-19   
4  8.914896e-16  2.484347e-16  8.562875e-16  7.157914e-14  3.332230e-18   

      sigma_per       rms  
0  0.000000e+00  0.000161  
1  1.921434e-16  0.000134  
2  1.621282e-16  0.000126  
3  6.623144e-17  0.000149  
4  1.821889e-16  0.000194

In [41]:
df10 = pd.concat((df7.reset_index(),df9.reset_index()), axis = 1)
df10.head()


index neo  orbit_id class  index         spkid         H     epoch  \
0      0   N    JPL 47   MBA      0  0.000000e+00  0.131195  0.988218   
1      1   N    JPL 37   MBA      1  1.922440e-08  0.154519  1.000000   
2      2   N  Other_id   MBA      2  3.844880e-08  0.187464  1.000000   
3      3   N    JPL 35   MBA      3  5.767320e-08  0.119534  0.988218   
4      4   N  Other_id   MBA      4  7.689760e-08  0.233236  1.000000   

   epoch_mjd  epoch_cal         e         a         q         i        om  \
0   0.988218   0.989134  0.076017  0.000066  0.030975  0.060467  0.223071   
1   1.000000   1.000000  0.230004  0.000066  0.025712  0.198916  0.480625   
2   1.000000   1.000000  0.256972  0.000063  0.023805  0.074158  0.471810   
3   0.988218   0.989134  0.088732  0.000054  0.025911  0.040748  0.288363   
4   1.000000   1.000000  0.190939  0.000060  0.025049  0.030614  0.393253   

          w        ma        ad         n        tp    tp_cal           per  \
0  0.204438  0.258627  0.000035  0.089827  0.665157  0.665258  6.843780e-07   
1  0.861674  0.379616  0.000041  0.089600  0.665470  0.665298  6.862835e-07   
2  0.689073  0.344645  0.000040  0.094969  0.665944  0.665354  6.436662e-07   
3  0.418691  0.291718  0.000029  0.114070  0.665191  0.665269  5.245750e-07   
4  0.996247  0.152094  0.000036  0.100232  0.667768  0.668020  6.063194e-07   

          per_y      moid   moid_ld       sigma_e       sigma_a       sigma_q  \
0  6.843780e-07  0.020066  0.020066  0.000000e+00  0.000000e+00  0.000000e+00   
1  6.862835e-07  0.015530  0.015530  8.098714e-13  1.241322e-15  8.699747e-14   
2  6.436662e-07  0.013014  0.013014  7.740059e-13  1.067960e-15  8.016988e-14   
3  5.245750e-07  0.014337  0.014337  5.793050e-15  4.640170e-16  1.880819e-15   
4  6.063194e-07  0.013787  0.013787  6.019577e-13  1.221946e-15  6.000437e-14   

        sigma_i      sigma_om       sigma_w      sigma_ma      sigma_ad  \
0  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
1  6.262048e-11  5.179478e-12  2.713159e-16  2.650639e-16  8.984495e-16   
2  5.816901e-11  1.383063e-11  5.285854e-16  2.426430e-16  7.900225e-16   
3  3.839709e-12  2.721975e-13  3.373713e-18  3.590511e-17  2.972008e-16   
4  4.945222e-11  2.409083e-11  8.914896e-16  2.484347e-16  8.562875e-16   

        sigma_n      sigma_tp     sigma_per       rms  
0  0.000000e+00  0.000000e+00  0.000000e+00  0.000161  
1  6.029037e-14  3.912546e-18  1.921434e-16  0.000134  
2  5.717040e-14  3.384558e-18  1.621282e-16  0.000126  
3  3.377845e-14  3.903861e-19  6.623144e-17  0.000149  
4  7.157914e-14  3.332230e-18  1.821889e-16  0.000194

In [42]:
df10.shape

(932335, 37)

In [43]:
df11 = df10.copy()

In [44]:
df11.head()

index neo  orbit_id class  index         spkid         H     epoch  \
0      0   N    JPL 47   MBA      0  0.000000e+00  0.131195  0.988218   
1      1   N    JPL 37   MBA      1  1.922440e-08  0.154519  1.000000   
2      2   N  Other_id   MBA      2  3.844880e-08  0.187464  1.000000   
3      3   N    JPL 35   MBA      3  5.767320e-08  0.119534  0.988218   
4      4   N  Other_id   MBA      4  7.689760e-08  0.233236  1.000000   

   epoch_mjd  epoch_cal         e         a         q         i        om  \
0   0.988218   0.989134  0.076017  0.000066  0.030975  0.060467  0.223071   
1   1.000000   1.000000  0.230004  0.000066  0.025712  0.198916  0.480625   
2   1.000000   1.000000  0.256972  0.000063  0.023805  0.074158  0.471810   
3   0.988218   0.989134  0.088732  0.000054  0.025911  0.040748  0.288363   
4   1.000000   1.000000  0.190939  0.000060  0.025049  0.030614  0.393253   

          w        ma        ad         n        tp    tp_cal           per  \
0  0.204438  0.258627  0.000035  0.089827  0.665157  0.665258  6.843780e-07   
1  0.861674  0.379616  0.000041  0.089600  0.665470  0.665298  6.862835e-07   
2  0.689073  0.344645  0.000040  0.094969  0.665944  0.665354  6.436662e-07   
3  0.418691  0.291718  0.000029  0.114070  0.665191  0.665269  5.245750e-07   
4  0.996247  0.152094  0.000036  0.100232  0.667768  0.668020  6.063194e-07   

          per_y      moid   moid_ld       sigma_e       sigma_a       sigma_q  \
0  6.843780e-07  0.020066  0.020066  0.000000e+00  0.000000e+00  0.000000e+00   
1  6.862835e-07  0.015530  0.015530  8.098714e-13  1.241322e-15  8.699747e-14   
2  6.436662e-07  0.013014  0.013014  7.740059e-13  1.067960e-15  8.016988e-14   
3  5.245750e-07  0.014337  0.014337  5.793050e-15  4.640170e-16  1.880819e-15   
4  6.063194e-07  0.013787  0.013787  6.019577e-13  1.221946e-15  6.000437e-14   

        sigma_i      sigma_om       sigma_w      sigma_ma      sigma_ad  \
0  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
1  6.262048e-11  5.179478e-12  2.713159e-16  2.650639e-16  8.984495e-16   
2  5.816901e-11  1.383063e-11  5.285854e-16  2.426430e-16  7.900225e-16   
3  3.839709e-12  2.721975e-13  3.373713e-18  3.590511e-17  2.972008e-16   
4  4.945222e-11  2.409083e-11  8.914896e-16  2.484347e-16  8.562875e-16   

        sigma_n      sigma_tp     sigma_per       rms  
0  0.000000e+00  0.000000e+00  0.000000e+00  0.000161  
1  6.029037e-14  3.912546e-18  1.921434e-16  0.000134  
2  5.717040e-14  3.384558e-18  1.621282e-16  0.000126  
3  3.377845e-14  3.903861e-19  6.623144e-17  0.000149  
4  7.157914e-14  3.332230e-18  1.821889e-16  0.000194

Categorical Encoding using get dummies for independent feature matrix in pandas

In [45]:
df12 = pd.get_dummies(df11, columns=['neo', 'class', 'orbit_id'])

In [46]:
df12.head()

index  index         spkid         H     epoch  epoch_mjd  epoch_cal  \
0      0      0  0.000000e+00  0.131195  0.988218   0.988218   0.989134   
1      1      1  1.922440e-08  0.154519  1.000000   1.000000   1.000000   
2      2      2  3.844880e-08  0.187464  1.000000   1.000000   1.000000   
3      3      3  5.767320e-08  0.119534  0.988218   0.988218   0.989134   
4      4      4  7.689760e-08  0.233236  1.000000   1.000000   1.000000   

          e         a         q         i        om         w        ma  \
0  0.076017  0.000066  0.030975  0.060467  0.223071  0.204438  0.258627   
1  0.230004  0.000066  0.025712  0.198916  0.480625  0.861674  0.379616   
2  0.256972  0.000063  0.023805  0.074158  0.471810  0.689073  0.344645   
3  0.088732  0.000054  0.025911  0.040748  0.288363  0.418691  0.291718   
4  0.190939  0.000060  0.025049  0.030614  0.393253  0.996247  0.152094   

         ad         n        tp    tp_cal           per         per_y  \
0  0.000035  0.089827  0.665157  0.665258  6.843780e-07  6.843780e-07   
1  0.000041  0.089600  0.665470  0.665298  6.862835e-07  6.862835e-07   
2  0.000040  0.094969  0.665944  0.665354  6.436662e-07  6.436662e-07   
3  0.000029  0.114070  0.665191  0.665269  5.245750e-07  5.245750e-07   
4  0.000036  0.100232  0.667768  0.668020  6.063194e-07  6.063194e-07   

       moid   moid_ld       sigma_e       sigma_a       sigma_q       sigma_i  \
0  0.020066  0.020066  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
1  0.015530  0.015530  8.098714e-13  1.241322e-15  8.699747e-14  6.262048e-11   
2  0.013014  0.013014  7.740059e-13  1.067960e-15  8.016988e-14  5.816901e-11   
3  0.014337  0.014337  5.793050e-15  4.640170e-16  1.880819e-15  3.839709e-12   
4  0.013787  0.013787  6.019577e-13  1.221946e-15  6.000437e-14  4.945222e-11   

       sigma_om       sigma_w      sigma_ma      sigma_ad       sigma_n  \
0  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
1  5.179478e-12  2.713159e-16  2.650639e-16  8.984495e-16  6.029037e-14   
2  1.383063e-11  5.285854e-16  2.426430e-16  7.900225e-16  5.717040e-14   
3  2.721975e-13  3.373713e-18  3.590511e-17  2.972008e-16  3.377845e-14   
4  2.409083e-11  8.914896e-16  2.484347e-16  8.562875e-16  7.157914e-14   

       sigma_tp     sigma_per       rms  neo_N  neo_Y  class_AMO  class_APO  \
0  0.000000e+00  0.000000e+00  0.000161      1      0          0          0   
1  3.912546e-18  1.921434e-16  0.000134      1      0          0          0   
2  3.384558e-18  1.621282e-16  0.000126      1      0          0          0   
3  3.903861e-19  6.623144e-17  0.000149      1      0          0          0   
4  3.332230e-18  1.821889e-16  0.000194      1      0          0          0   

   class_AST  class_ATE  class_CEN  class_IEO  class_IMB  class_MBA  \
0          0          0          0          0          0          1   
1          0          0          0          0          0          1   
2          0          0          0          0          0          1   
3          0          0          0          0          0          1   
4          0          0          0          0          0          1   

   class_MCA  class_OMB  class_TJN  class_TNO  orbit_id_1  orbit_id_10  \
0          0          0          0          0           0            0   
1          0          0          0          0           0            0   
2          0          0          0          0           0            0   
3          0          0          0          0           0            0   
4          0          0          0          0           0            0   

   orbit_id_101  orbit_id_103  orbit_id_104  orbit_id_11  orbit_id_12  \
0             0             0             0            0            0   
1             0             0             0            0            0   
2             0             0             0            0            0   
3             0             0             0            0            0   
4           

Label encodig of Dependent Variable 

In [47]:
from sklearn.preprocessing import LabelEncoder
le = LabelEncoder()
df8 = le.fit_transform(df8)
df8

array([0, 0, 0, ..., 1, 0, 0])

In [48]:
list(le.classes_)

['N', 'Y']

In [49]:
df13 = df12.drop(['index', 'index'], axis = 'columns')

In [50]:
df14 = df13.set_index('spkid')

In [51]:
df14.head()

H     epoch  epoch_mjd  epoch_cal         e         a  \
spkid                                                                        
0.000000e+00  0.131195  0.988218   0.988218   0.989134  0.076017  0.000066   
1.922440e-08  0.154519  1.000000   1.000000   1.000000  0.230004  0.000066   
3.844880e-08  0.187464  1.000000   1.000000   1.000000  0.256972  0.000063   
5.767320e-08  0.119534  0.988218   0.988218   0.989134  0.088732  0.000054   
7.689760e-08  0.233236  1.000000   1.000000   1.000000  0.190939  0.000060   

                     q         i        om         w        ma        ad  \
spkid                                                                      
0.000000e+00  0.030975  0.060467  0.223071  0.204438  0.258627  0.000035   
1.922440e-08  0.025712  0.198916  0.480625  0.861674  0.379616  0.000041   
3.844880e-08  0.023805  0.074158  0.471810  0.689073  0.344645  0.000040   
5.767320e-08  0.025911  0.040748  0.288363  0.418691  0.291718  0.000029   
7.689760e-08  0.025049  0.030614  0.393253  0.996247  0.152094  0.000036   

                     n        tp    tp_cal           per         per_y  \
spkid                                                                    
0.000000e+00  0.089827  0.665157  0.665258  6.843780e-07  6.843780e-07   
1.922440e-08  0.089600  0.665470  0.665298  6.862835e-07  6.862835e-07   
3.844880e-08  0.094969  0.665944  0.665354  6.436662e-07  6.436662e-07   
5.767320e-08  0.114070  0.665191  0.665269  5.245750e-07  5.245750e-07   
7.689760e-08  0.100232  0.667768  0.668020  6.063194e-07  6.063194e-07   

                  moid   moid_ld       sigma_e       sigma_a       sigma_q  \
spkid                                                                        
0.000000e+00  0.020066  0.020066  0.000000e+00  0.000000e+00  0.000000e+00   
1.922440e-08  0.015530  0.015530  8.098714e-13  1.241322e-15  8.699747e-14   
3.844880e-08  0.013014  0.013014  7.740059e-13  1.067960e-15  8.016988e-14   
5.767320e-08  0.014337  0.014337  5.793050e-15  4.640170e-16  1.880819e-15   
7.689760e-08  0.013787  0.013787  6.019577e-13  1.221946e-15  6.000437e-14   

                   sigma_i      sigma_om       sigma_w      sigma_ma  \
spkid                                                                  
0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
1.922440e-08  6.262048e-11  5.179478e-12  2.713159e-16  2.650639e-16   
3.844880e-08  5.816901e-11  1.383063e-11  5.285854e-16  2.426430e-16   
5.767320e-08  3.839709e-12  2.721975e-13  3.373713e-18  3.590511e-17   
7.689760e-08  4.945222e-11  2.409083e-11  8.914896e-16  2.484347e-16   

                  sigma_ad       sigma_n      sigma_tp     sigma_per  \
spkid                                                                  
0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
1.922440e-08  8.984495e-16  6.029037e-14  3.912546e-18  1.921434e-16   
3.844880e-08  7.900225e-16  5.717040e-14  3.384558e-18  1.621282e-16   
5.767320e-08  2.972008e-16  3.377845e-14  3.903861e-19  6.623144e-17   
7.689760e-08  8.562875e-16  7.157914e-14  3.332230e-18  1.821889e-16   

                   rms  neo_N  neo_Y  class_AMO  class_APO  class_AST  \
spkid                                                                   
0.000000e+00  0.000161      1      0          0          0          0   
1.922440e-08  0.000134      1      0          0          0          0   
3.844880e-08  0.000126      1      0          0          0          0   
5.767320e-08  0.000149      1      0          0          0          0   
7.689760e-08  0.000194      1      0          0          0          0   

              class_ATE  class_CEN  class_IEO  class_IMB  class_MBA  \
spkid                                                                 
0.000000e+00          0          0          0          0          1   
1.922440e-08          0          0          0          0          1   
3.844880e-08          0          0          0          0          1  

# Model Building

In [52]:
from sklearn.model_selection import train_test_split
X = df13
y = df8
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=0)

The data is highly imbalanced with over 99% of the data belonging to the negative class. This could sway the models and predict only the negative class for any input. For this reason, its best to oversample the positive class and create an equal sample numbers for both classes. This is achieved by usig the library SMOTE.

In [53]:
#Before Oversampling the number of N's and Y's
print('Before oversampling the number of N : {}'.format(sum(y_train == 0 )))
print('Before oversampling the number of Y : {}'.format(sum(y_train == 1)))

# import SMOTE module from imblearn library 
from imblearn.over_sampling import SMOTE 
sm = SMOTE(random_state = 12) 
X_train_res, y_train_res = sm.fit_sample(X_train, y_train.ravel()) 

print('After oversampling the number of N : {}'.format(sum(y_train_res == 0)))
print('After oversampling the number of Y : {}'.format(sum(y_train_res == 1)))


Before oversampling the number of N : 697705
Before oversampling the number of Y : 1546
After oversampling the number of N : 697705
After oversampling the number of Y : 697705


In [54]:
#Writing a function to find metric of each model
def metricCalculation(y_test, pred):
    
    precision_metric = metrics.precision_score(y_test, pred)
    recall_metric = metrics.recall_score(y_test, pred)
    accuracy_metric = metrics.accuracy_score(y_test, pred)
    f1_metric = metrics.f1_score(y_test, pred)
    print('Precision metric:',round(precision_metric, 2))
    print('Recall Metric:',round(recall_metric, 2))
    print('Accuracy Metric:',round(accuracy_metric, 4))
    print('F1 score:',round(f1_metric, 2))

# Logistic Regression Model

Logistic Regression will be the baseline model for the dataset. Using the metrics from this model, we can compare metrics from the other models and tune them to achieve better values.

In [55]:
#Importing the Logistic Regression Model
from sklearn.linear_model import LogisticRegression

#Creating an object of this model
classifier = LogisticRegression(random_state = 0, max_iter= 10000)

#Fitting the Model
classifier.fit(X_train_res, y_train_res)

LogisticRegression(max_iter=10000, random_state=0)

In [56]:
#Predicting the result for the test set
lr_pred = classifier.predict(X_test)

In [57]:
#Printing metrics
metricCalculation(y_test, lr_pred)

Precision metric: 0.19
Recall Metric: 0.98
Accuracy Metric: 0.9907
F1 score: 0.32


In [58]:
#Confusion Matrix

metrics.confusion_matrix(y_test, lr_pred)

array([[230401,   2163],
       [     9,    511]], dtype=int64)

# Random Forest 

In [59]:
#Importing Random Forest Module
from sklearn.ensemble import RandomForestClassifier

#Creating an object of Random Forest Model
classifier = RandomForestClassifier(n_estimators=100, criterion='entropy', random_state = 0)

#Fitting the Random Forest model
classifier.fit(X_train_res, y_train_res)


RandomForestClassifier(criterion='entropy', random_state=0)

In [60]:
#Predicting the result for test set
rf_pred = classifier.predict(X_test) 

In [61]:
#Printing metrics
metricCalculation(y_test, rf_pred)

Precision metric: 0.98
Recall Metric: 0.98
Accuracy Metric: 0.9999
F1 score: 0.98


In [62]:
#Confusion Matrix

metrics.confusion_matrix(y_test, rf_pred)

array([[232553,     11],
       [     8,    512]], dtype=int64)

 We can identify the most important features that help in determining the type of asteroid.

In [63]:
feature_imp = pd.DataFrame(classifier.feature_importances_,index=X_train_res.columns, columns = ['Importance']).sort_values(by='Importance', ascending=False)

In [64]:
feature_imp

Importance
moid_ld            0.177900
moid               0.147549
q                  0.122472
class_MBA          0.100434
neo_N              0.091095
...                     ...
sigma_w            0.000000
class_AST          0.000000
orbit_id_JPL 81    0.000000
sigma_tp           0.000000
orbit_id_JPL 68    0.000000

[241 rows x 1 columns]

In [65]:
# Top 10 important variables
feature_imp[0:10]

Importance
moid_ld      0.177900
moid         0.147549
q            0.122472
class_MBA    0.100434
neo_N        0.091095
neo_Y        0.082686
class_APO    0.070884
e            0.067278
H            0.045379
n            0.031580

In [66]:
#Least 20 important variables
feature_imp[:-50]

Importance
moid_ld          1.779002e-01
moid             1.475486e-01
q                1.224723e-01
class_MBA        1.004341e-01
neo_N            9.109534e-02
...                       ...
orbit_id_JPL 60  4.361375e-07
orbit_id_JPL 43  4.244531e-07
orbit_id_JPL 57  4.131691e-07
orbit_id_JPL 77  4.064907e-07
orbit_id_JPL 49  4.011355e-07

[191 rows x 1 columns]

Upon further exploring the least important features, it can be seen that the orbit IDs do not contribute much to the model. The dataset can be modified by eliminating the orbit_id feature completely. Similarly we can also elimate the features 'sigma_ma' and 'sigma_tp' that have 0 importance.

To do this, create a new dataset with one-hot encoding and dropping the orbit_id column.

In [67]:
df11.head()

index neo  orbit_id class  index         spkid         H     epoch  \
0      0   N    JPL 47   MBA      0  0.000000e+00  0.131195  0.988218   
1      1   N    JPL 37   MBA      1  1.922440e-08  0.154519  1.000000   
2      2   N  Other_id   MBA      2  3.844880e-08  0.187464  1.000000   
3      3   N    JPL 35   MBA      3  5.767320e-08  0.119534  0.988218   
4      4   N  Other_id   MBA      4  7.689760e-08  0.233236  1.000000   

   epoch_mjd  epoch_cal         e         a         q         i        om  \
0   0.988218   0.989134  0.076017  0.000066  0.030975  0.060467  0.223071   
1   1.000000   1.000000  0.230004  0.000066  0.025712  0.198916  0.480625   
2   1.000000   1.000000  0.256972  0.000063  0.023805  0.074158  0.471810   
3   0.988218   0.989134  0.088732  0.000054  0.025911  0.040748  0.288363   
4   1.000000   1.000000  0.190939  0.000060  0.025049  0.030614  0.393253   

          w        ma        ad         n        tp    tp_cal           per  \
0  0.204438  0.258627  0.000035  0.089827  0.665157  0.665258  6.843780e-07   
1  0.861674  0.379616  0.000041  0.089600  0.665470  0.665298  6.862835e-07   
2  0.689073  0.344645  0.000040  0.094969  0.665944  0.665354  6.436662e-07   
3  0.418691  0.291718  0.000029  0.114070  0.665191  0.665269  5.245750e-07   
4  0.996247  0.152094  0.000036  0.100232  0.667768  0.668020  6.063194e-07   

          per_y      moid   moid_ld       sigma_e       sigma_a       sigma_q  \
0  6.843780e-07  0.020066  0.020066  0.000000e+00  0.000000e+00  0.000000e+00   
1  6.862835e-07  0.015530  0.015530  8.098714e-13  1.241322e-15  8.699747e-14   
2  6.436662e-07  0.013014  0.013014  7.740059e-13  1.067960e-15  8.016988e-14   
3  5.245750e-07  0.014337  0.014337  5.793050e-15  4.640170e-16  1.880819e-15   
4  6.063194e-07  0.013787  0.013787  6.019577e-13  1.221946e-15  6.000437e-14   

        sigma_i      sigma_om       sigma_w      sigma_ma      sigma_ad  \
0  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
1  6.262048e-11  5.179478e-12  2.713159e-16  2.650639e-16  8.984495e-16   
2  5.816901e-11  1.383063e-11  5.285854e-16  2.426430e-16  7.900225e-16   
3  3.839709e-12  2.721975e-13  3.373713e-18  3.590511e-17  2.972008e-16   
4  4.945222e-11  2.409083e-11  8.914896e-16  2.484347e-16  8.562875e-16   

        sigma_n      sigma_tp     sigma_per       rms  
0  0.000000e+00  0.000000e+00  0.000000e+00  0.000161  
1  6.029037e-14  3.912546e-18  1.921434e-16  0.000134  
2  5.717040e-14  3.384558e-18  1.621282e-16  0.000126  
3  3.377845e-14  3.903861e-19  6.623144e-17  0.000149  
4  7.157914e-14  3.332230e-18  1.821889e-16  0.000194

In [68]:
df14 = df11.drop(['index', 'orbit_id', 'index', 'spkid', 'sigma_ma', 'sigma_tp' ], axis = 'columns')

In [69]:
df14.head()

neo class         H     epoch  epoch_mjd  epoch_cal         e         a  \
0   N   MBA  0.131195  0.988218   0.988218   0.989134  0.076017  0.000066   
1   N   MBA  0.154519  1.000000   1.000000   1.000000  0.230004  0.000066   
2   N   MBA  0.187464  1.000000   1.000000   1.000000  0.256972  0.000063   
3   N   MBA  0.119534  0.988218   0.988218   0.989134  0.088732  0.000054   
4   N   MBA  0.233236  1.000000   1.000000   1.000000  0.190939  0.000060   

          q         i        om         w        ma        ad         n  \
0  0.030975  0.060467  0.223071  0.204438  0.258627  0.000035  0.089827   
1  0.025712  0.198916  0.480625  0.861674  0.379616  0.000041  0.089600   
2  0.023805  0.074158  0.471810  0.689073  0.344645  0.000040  0.094969   
3  0.025911  0.040748  0.288363  0.418691  0.291718  0.000029  0.114070   
4  0.025049  0.030614  0.393253  0.996247  0.152094  0.000036  0.100232   

         tp    tp_cal           per         per_y      moid   moid_ld  \
0  0.665157  0.665258  6.843780e-07  6.843780e-07  0.020066  0.020066   
1  0.665470  0.665298  6.862835e-07  6.862835e-07  0.015530  0.015530   
2  0.665944  0.665354  6.436662e-07  6.436662e-07  0.013014  0.013014   
3  0.665191  0.665269  5.245750e-07  5.245750e-07  0.014337  0.014337   
4  0.667768  0.668020  6.063194e-07  6.063194e-07  0.013787  0.013787   

        sigma_e       sigma_a       sigma_q       sigma_i      sigma_om  \
0  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
1  8.098714e-13  1.241322e-15  8.699747e-14  6.262048e-11  5.179478e-12   
2  7.740059e-13  1.067960e-15  8.016988e-14  5.816901e-11  1.383063e-11   
3  5.793050e-15  4.640170e-16  1.880819e-15  3.839709e-12  2.721975e-13   
4  6.019577e-13  1.221946e-15  6.000437e-14  4.945222e-11  2.409083e-11   

        sigma_w      sigma_ad       sigma_n     sigma_per       rms  
0  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000161  
1  2.713159e-16  8.984495e-16  6.029037e-14  1.921434e-16  0.000134  
2  5.285854e-16  7.900225e-16  5.717040e-14  1.621282e-16  0.000126  
3  3.373713e-18  2.972008e-16  3.377845e-14  6.623144e-17  0.000149  
4  8.914896e-16  8.562875e-16  7.157914e-14  1.821889e-16  0.000194

In [70]:
df15 = pd.get_dummies(df14, columns=['neo', 'class'])

In [71]:
df15.head()

H     epoch  epoch_mjd  epoch_cal         e         a         q  \
0  0.131195  0.988218   0.988218   0.989134  0.076017  0.000066  0.030975   
1  0.154519  1.000000   1.000000   1.000000  0.230004  0.000066  0.025712   
2  0.187464  1.000000   1.000000   1.000000  0.256972  0.000063  0.023805   
3  0.119534  0.988218   0.988218   0.989134  0.088732  0.000054  0.025911   
4  0.233236  1.000000   1.000000   1.000000  0.190939  0.000060  0.025049   

          i        om         w        ma        ad         n        tp  \
0  0.060467  0.223071  0.204438  0.258627  0.000035  0.089827  0.665157   
1  0.198916  0.480625  0.861674  0.379616  0.000041  0.089600  0.665470   
2  0.074158  0.471810  0.689073  0.344645  0.000040  0.094969  0.665944   
3  0.040748  0.288363  0.418691  0.291718  0.000029  0.114070  0.665191   
4  0.030614  0.393253  0.996247  0.152094  0.000036  0.100232  0.667768   

     tp_cal           per         per_y      moid   moid_ld       sigma_e  \
0  0.665258  6.843780e-07  6.843780e-07  0.020066  0.020066  0.000000e+00   
1  0.665298  6.862835e-07  6.862835e-07  0.015530  0.015530  8.098714e-13   
2  0.665354  6.436662e-07  6.436662e-07  0.013014  0.013014  7.740059e-13   
3  0.665269  5.245750e-07  5.245750e-07  0.014337  0.014337  5.793050e-15   
4  0.668020  6.063194e-07  6.063194e-07  0.013787  0.013787  6.019577e-13   

        sigma_a       sigma_q       sigma_i      sigma_om       sigma_w  \
0  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
1  1.241322e-15  8.699747e-14  6.262048e-11  5.179478e-12  2.713159e-16   
2  1.067960e-15  8.016988e-14  5.816901e-11  1.383063e-11  5.285854e-16   
3  4.640170e-16  1.880819e-15  3.839709e-12  2.721975e-13  3.373713e-18   
4  1.221946e-15  6.000437e-14  4.945222e-11  2.409083e-11  8.914896e-16   

       sigma_ad       sigma_n     sigma_per       rms  neo_N  neo_Y  \
0  0.000000e+00  0.000000e+00  0.000000e+00  0.000161      1      0   
1  8.984495e-16  6.029037e-14  1.921434e-16  0.000134      1      0   
2  7.900225e-16  5.717040e-14  1.621282e-16  0.000126      1      0   
3  2.972008e-16  3.377845e-14  6.623144e-17  0.000149      1      0   
4  8.562875e-16  7.157914e-14  1.821889e-16  0.000194      1      0   

   class_AMO  class_APO  class_AST  class_ATE  class_CEN  class_IEO  \
0          0          0          0          0          0          0   
1          0          0          0          0          0          0   
2          0          0          0          0          0          0   
3          0          0          0          0          0          0   
4          0          0          0          0          0          0   

   class_IMB  class_MBA  class_MCA  class_OMB  class_TJN  class_TNO  
0          0          1          0          0          0          0  
1          0          1          0          0          0          0  
2          0          1          0          0          0          0  
3          0          1          0          0          0          0  
4          0          1          0          0          0          0

# Random Forest trained on trimmed data

In [72]:
from sklearn.model_selection import train_test_split
X1 = df15
y1 = df8
X1_train, X1_test, y1_train, y1_test = train_test_split(X1, y1, test_size=0.25, random_state=0)

In [73]:
#Before Oversampling the number of N's and Y's
print('Before oversampling the number of N : {}'.format(sum(y_train == 0 )))
print('Before oversampling the number of Y : {}'.format(sum(y_train == 1)))

# import SMOTE module from imblearn library 
from imblearn.over_sampling import SMOTE 
sm = SMOTE(random_state = 12) 
X1_train_res, y1_train_res = sm.fit_sample(X1_train, y1_train.ravel()) 

print('After oversampling the number of N : {}'.format(sum(y_train_res == 0)))
print('After oversampling the number of Y : {}'.format(sum(y_train_res == 1)))


Before oversampling the number of N : 697705
Before oversampling the number of Y : 1546
After oversampling the number of N : 697705
After oversampling the number of Y : 697705


In [74]:
#Importing Random Forest Module
from sklearn.ensemble import RandomForestClassifier

#Creating an object of Random Forest Model
rf2_classifier = RandomForestClassifier(n_estimators=100, criterion='entropy', random_state = 0)

#Fitting the Random Forest model
rf2_classifier.fit(X1_train_res, y1_train_res)

RandomForestClassifier(criterion='entropy', random_state=0)

In [75]:
rf2_pred = rf2_classifier.predict(X1_test)

In [76]:
metricCalculation(y1_test, rf2_pred)

Precision metric: 0.98
Recall Metric: 0.99
Accuracy Metric: 0.9999
F1 score: 0.98


In [77]:
metrics.confusion_matrix(y1_test, rf2_pred)

array([[232552,     12],
       [     7,    513]], dtype=int64)

As can be seen, the model performance has improved with fewer false positives and false negatives. Thus, Random Forest can be used as a reliable model to predict the nature of the asteroid.

# 3.Light Gradient Boosting

Gradient Boosting Model (GBM) is a model better preferred for prediction since it combines the principles of gradient decsent and the randomness of decision trees. We can expect a better performing model with GBM, as compared to Random Forest. Due to the massive size of the data, we can choose Light Gradient Boosting model which is known for its quick performance when compared to XGBoosting.

This algorithm will be used on 2 datasets - one with all the features and one without the important features as identified by the Random Forest model

# 3.1 Light Gradient Boosting trained on entired data set

In [78]:
#Load the training dataset along with the label to LGBM
import lightgbm as lgb 

train_data=lgb.Dataset(X_train_res,label=y_train_res)

Set paramateres before fitting the model. After experimenting with a few learning rates, it was found that a rate of 0.01 yeilded the highest value for precision although all other metrics remained almost the same.

In [79]:
#setting parameters for lightgbm

param = {'num_leaves': 150, # number of leaves per tree
         'nrounds': 350,
         'max_depth': 25, # depth of tree
         'learning_rate': 0.01, # learning rate
         'max_bin': 500 # max number of bins to bucket the feature values.
        }

In [80]:
# Train the model 

lgbm = lgb.train(param, train_data)
lgbm_pred = lgbm.predict(X_test)

# Convert the predicted probabilities to 0 or 1
for i in range(0,len(y_test)):
    if lgbm_pred[i]>=.5:       # setting threshold to .5
       lgbm_pred[i]=1
    else:  
       lgbm_pred[i]=0

[LightGBM] [Warning] Unknown parameter: nrounds
[LightGBM] [Warning] Found whitespace in feature_names, replace with underlines
[LightGBM] [Warning] Unknown parameter: nrounds
[LightGBM] [Warning] Auto-choosing col-wise multi-threading, the overhead of testing was 0.301399 seconds.
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info] Total Bins 16370
[LightGBM] [Info] Number of data points in the train set: 1395410, number of used features: 217
[LightGBM] [Info] Start training from score 0.500000
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No f

In [81]:
# Calculate metrics
metricCalculation(y_test, lgbm_pred)

Precision metric: 0.96
Recall Metric: 0.98
Accuracy Metric: 0.9998
F1 score: 0.97


In [82]:
# Confusion Matrix
print(metrics.confusion_matrix(y_test, lgbm_pred))

[[232541     23]
 [    12    508]]


The Light Gradient Boosting Model with the entire dataset has higher metric values as compared to Logistic Regression but a little lower than Random Forest. The confusion matrix shows that the LGBM has more mislabled values than the Random Forest model. Now we try training the model on the trimmed data set based on the importance of the Random Forest model.

# 3.2 Light Gradient Boosting trained on trimmed data from random forest

In [83]:
#Load the training dataset along with the label to LGBM
import lightgbm as lgb2 

train_data1=lgb2.Dataset(X1_train_res,label=y1_train_res)

In [84]:
#setting parameters for lightgbm

param = {'num_leaves': 150, # number of leaves per tree
         'nrounds': 350,
         'max_depth': 25, # depth of tree
         'learning_rate': 0.01, # learning rate
         'max_bin': 500 # max number of bins to bucket the feature values.
        }

In [85]:
# Train the model 

lgbm1 = lgb.train(param, train_data1)
lgbm_pred1 = lgbm1.predict(X1_test)

# Convert the predicted probabilities to 0 or 1
for i in range(0,len(y1_test)):
    if lgbm_pred1[i]>=.5:       # setting threshold to .5
       lgbm_pred1[i]=1
    else:  
       lgbm_pred1[i]=0

[LightGBM] [Warning] Unknown parameter: nrounds
[LightGBM] [Warning] Unknown parameter: nrounds
[LightGBM] [Warning] Auto-choosing col-wise multi-threading, the overhead of testing was 0.378734 seconds.
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info] Total Bins 14528
[LightGBM] [Info] Number of data points in the train set: 1395410, number of used features: 43
[LightGBM] [Info] Start training from score 0.500000
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further

In [86]:
# Calculate metrics
metricCalculation(y_test, lgbm_pred1)

Precision metric: 0.96
Recall Metric: 0.98
Accuracy Metric: 0.9999
F1 score: 0.97


In [87]:
# Confusion Matrix
print(metrics.confusion_matrix(y_test, lgbm_pred))

[[232541     23]
 [    12    508]]


# Conclusion

| Model| Accuracy  | Precision    | Recall   | F-1 Score   |
|---:|:-------------|:-----------|:------|:------|
| Random Forest with importance | 99.99%  | 0.98    | 0.99   | 0.98     |
| Random Forest | 99.99%  | 0.98   | 0.98   | 0.98     |
| Light Gradient Boosting with importance | 99.99%  | 0.9999    | 0.98   | 0.97     |
| Light Gradient Boosting | 99.98%  |  0.96  |  0.98  |   0.97  |
| Logistic Regression |  99.07% | 0.19       | 0.98   | 0.32     |

The Random Forest model with only important features trumps the other models in perfomance metrics. Light Gradient Boosting had a good performance as well, but not as good as Random Forest, even with the model tuned for different paramters. Logistic Regression was used a baseline model to match the other models with, and although it had good accuracy, it still was weak in performance.

Thus a tuned Random Forest model would be best to predict the hazardous nature of the asteroids.